### ISSUE TITLE
[1.5.0-RC3] filters: undefined local variable or method `event` 

### ISSUE THREAD

**Issue by suyograo** [elastic/logstash#3100]
As reported by @wjimenez5271 using 1.5.0-RC3 while debugging https://github.com/elastic/logstash/issues/3092

```
$ sudo /opt/logstash/bin/logstash -f /etc/logstash.d/lumberjack.conf
Exception in filterworker {"exception"=>#<NameError: undefined local variable or method `event' for #<LogStash::Pipeline:0x51a97faa>>, "backtrace"=>["(eval):111:in `initialize'", "org/jruby/RubyProc.java:271:in `call'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.rc3-java/lib/logstash/pipeline.rb:309:in `flush_filters'", "org/jruby/RubyArray.java:1613:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.rc3-java/lib/logstash/pipeline.rb:308:in `flush_filters'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.rc3-java/lib/logstash/pipeline.rb:317:in `flush_filters_to_output!'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.rc3-java/lib/logstash/pipeline.rb:223:in `filterworker'", "org/jruby/ext/thread/Mutex.java:149:in `synchronize'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.rc3-java/lib/logstash/pipeline.rb:223:in `filterworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.rc3-java/lib/logstash/pipeline.rb:156:in `start_filters'"], :level=>:error}
Logstash startup completed
cx^CSIGINT received. Shutting down the pipeline. {:level=>:warn}
Exception in lumberjack input {:exception=>#<LogStash::ShutdownSignal: LogStash::ShutdownSignal>, :level=>:error}
```

Config: https://gist.github.com/wjimenez5271/d145cf8906752b5a5cbe

**Comment by jordansissel** [comment]
> "backtrace"=>["(eval):111:in `initialize'", "org/jruby/RubyProc.java:271:in `call'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0.rc3-java/lib/logstash/pipeline.rb:309:in `flush_filters'"

A crash in eval? Probably a bug in the config compiler that is specific to filter flushing.

**Comment by wjimenez5271** [comment]
Example log line:
```
2015-04-23 00:00:18,452 DEBUG [class] Log Text
```

Java details:
```
java version "1.8.0_31"
Java(TM) SE Runtime Environment (build 1.8.0_31-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.31-b07, mixed mode)
```

**Comment by suyograo** [comment]
@jordansissel we did change the filter code in https://github.com/elastic/logstash/pull/2869

**Comment by wjimenez5271** [comment]
@jordansissel Here is the debug output for the config portion: https://gist.github.com/wjimenez5271/17b92b1ae303b04d993f

**Comment by wjimenez5271** [comment]
And just to confirm we're using 0.4.0 of logstash forwarder

**Comment by jordansissel** [comment]
I used this config: https://gist.github.com/wjimenez5271/d145cf8906752b5a5cbe (provided in the description of this ticket)

I am unable to reproduce this problem. I run logstash and it starts up and idles (no events because I haven't setup logstash-forwarder). I might need to ship some sample events to see what's going on.

**Comment by wjimenez5271** [comment]
That would make sense, I see `Logstash startup completed` before the aforementioned exception.

**Comment by wjimenez5271** [comment]
@jordansissel I'm not sure what to provide you in terms of example events, we have a large variety of them hitting this lumberjack instance. The example I mentioned above is representative of the format of at least half of them. Are there any probable suspects I can look for in the event stream?

**Comment by dougmcclure** [comment]
I got something very similar with my first RC3 attempt after using RC2 for
some time.  Not using lumberjack, just TCP and File inputs here.  I ended
up reverting back to RC2.

**Comment by jordansissel** [comment]
@dougmcclure that helps, thank you! :)

**Comment by wjimenez5271** [comment]
OK I'll roll back to RC2 in the meantime. Sorry I couldn't be more helpful.

**Comment by suyograo** [comment]
@dougmcclure thanks, could you please provide us the config you used and the generated config by using `bin/logstash --debug`

**Comment by wiibaa** [comment]
I could achieve the same exception with this minimal config, sending an event that get hold until flushing, if this may help you understand what went wrong
````
input {
	stdin {}
}
filter {
  multiline {
   pattern => "hello"
   what => next
  }
  multiline {
   pattern => "hello"
   what => next
  }
}
output {
	stdout {}
}
````
debug output is here https://gist.github.com/wiibaa/96b2412f265d4266455f

**Comment by jordansissel** [comment]
https://gist.github.com/wiibaa/96b2412f265d4266455f#file-gistfile1-txt-L28 is the bug.

Caused by https://github.com/elastic/logstash/blob/master/lib/logstash/config/config_ast.rb#L233

This should be easy to fix.

**Comment by jordansissel** [comment]
Proposed fix with my eyeballs:

https://github.com/elastic/logstash/blob/master/lib/logstash/config/config_ast.rb#L158

Wrap this in a loop over `events`:

```ruby
events.each do |event|
  #{plugin.compile_starting_here.gsub(/^/, "  ")}
end
```

**Comment by suyograo** [comment]
Thanks much @wiibaa, you rock!

**Comment by jordansissel** [comment]
Reproducing w/ @wiibaa's config. I typed "hello" and waited for flushing to go through:

```
% bin/logstash -f conf
Logstash startup completed
hello
Exception in filterworker {"exception"=>#<NameError: undefined local variable or method `event' for #<LogStash::Pipeline:0x1a8cbd5>>, "backtrace"=>["(eval):24:in `initialize'", "org/jruby/RubyProc.java:271:in `call'", "/Users/jls/projects/logstash/lib/logstash/pipeline.rb:309:in `flush_filters'", "org/jruby/RubyArray.java:1613:in `each'", "/Users/jls/projects/logstash/lib/logstash/pipeline.rb:308:in `flush_filters'", "/Users/jls/projects/logstash/lib/logstash/pipeline.rb:317:in `flush_filters_to_output!'", "/Users/jls/projects/logstash/lib/logstash/pipeline.rb:223:in `filterworker'", "org/jruby/ext/thread/Mutex.java:149:in `synchronize'", "/Users/jls/projects/logstash/lib/logstash/pipeline.rb:223:in `filterworker'", "/Users/jls/projects/logstash/lib/logstash/pipeline.rb:156:in `start_filters'"], :level=>:error}
^CSIGINT received. Shutting down the pipeline. {:level=>:warn}
2015-04-24T20:27:48.018Z Macintosh-3.lan hello
Logstash shutdown completed
LogStash::ShutdownSignal: LogStash::ShutdownSignal
       debug? at /Users/jls/projects/logstash/vendor/bundle/jruby/1.9/gems/cabin-0.7.1/lib/cabin/mixins/logger.rb:70
  inputworker at /Users/jls/projects/logstash/lib/logstash/pipeline.rb:181
  start_input at /Users/jls/projects/logstash/lib/logstash/pipeline.rb:171
```

**Comment by jordansissel** [comment]
```diff
-              #{plugin.compile_starting_here.gsub(/^/, "  ")}
+              events.each do |event|
+                #{plugin.compile_starting_here.gsub(/^/, "  ")}
+              end
```

Result:

```
% bin/logstash -f conf
Logstash startup completed
hello
2015-04-24T20:30:16.552Z Macintosh-3.lan hello
hello
```

No crash, but looks like two "hello" lines in one event? I change to using `rubydebug` codec on stdout and get this:

```
% bin/logstash -f conf
Logstash startup completed
hello
{
  "message"    => "hello\nhello",
  "@version"   => "1",
  "@timestamp" => "2015-04-24T20:32:50.756Z",
  "host"       => "Macintosh-3.lan",
  "tags"       => [
    [0] "multiline"
  ]
}
```

I think this is a bug, maybe in my patch, not sure.

**Comment by jordansissel** [comment]
^^ Previous comment. The second bug appears to be that event cancellation is ignored.

I'm still digging.

**Comment by jordansissel** [comment]
I've reproduced this bug sufficiently that I thinK I know the conditions when it occurs: This bug is triggered whenever a filter that uses flush (metrics or multiline) occurs in the config and any other filter occurs after it.

I can reproduce it with @wiibaa's config (multiline + multiline) and also my config (multiline + drop).

**Comment by jordansissel** [comment]
I've got a tentative patch here: https://github.com/elastic/logstash/compare/1.5...jordansissel:bug/3100

You can apply it locally with `curl -s https://github.com/elastic/logstash/compare/1.5...jordansissel:bug/3100.diff | patch -p1` from the root of your logstash git tree.

The patch fixes two bugs.

**Comment by colinsurprenant** [comment]
I also submitted #3110 for the work I did on this while on the plane from Berlin :P @jordansissel please take a look and let me know what you think..

**Comment by colinsurprenant** [comment]
fixed by #3110 - thanks @wjimenez5271 @dougmcclure @wiibaa @jordansissel for the troubleshooting help on this!!
